### PR TITLE
LSM: Use `TableImmutable.Iterator` for Compaction

### PR DESCRIPTION
**Pull request by kprotty**
This allows for future implementations of TabelImmutable to not require providing a contiguous array of sorted Values. A `memcpy` is still used when arrays are detected.

## Pre-merge checklist

Performance:

* [x] Compare `zig build benchmark` on linux before and after this pr.
    ``` sh
    # benchmark results before
    1224 batches in 87.15 s
    load offered = 1000000 tx/s
    load accepted = 114740 tx/s
    batch latency p00 = 0 ms
    batch latency p10 = 28 ms
    batch latency p20 = 28 ms
    batch latency p30 = 29 ms
    batch latency p40 = 30 ms
    batch latency p50 = 31 ms
    batch latency p60 = 32 ms
    batch latency p70 = 32 ms
    batch latency p80 = 33 ms
    batch latency p90 = 34 ms
    batch latency p100 = 4445 ms
    transfer latency p00 = 0 ms
    transfer latency p10 = 3915 ms
    transfer latency p20 = 8092 ms
    transfer latency p30 = 14546 ms
    transfer latency p40 = 22390 ms
    transfer latency p50 = 31734 ms
    transfer latency p60 = 41478 ms
    transfer latency p70 = 51267 ms
    transfer latency p80 = 61130 ms
    transfer latency p90 = 70222 ms
    transfer latency p100 = 77168 ms
    
    # benchmark results after
    1224 batches in 86.30 s
    load offered = 1000000 tx/s
    load accepted = 115873 tx/s
    batch latency p00 = 0 ms
    batch latency p10 = 27 ms
    batch latency p20 = 28 ms
    batch latency p30 = 29 ms
    batch latency p40 = 29 ms
    batch latency p50 = 30 ms
    batch latency p60 = 31 ms
    batch latency p70 = 32 ms
    batch latency p80 = 32 ms
    batch latency p90 = 33 ms
    batch latency p100 = 4279 ms
    transfer latency p00 = 0 ms
    transfer latency p10 = 3893 ms
    transfer latency p20 = 7994 ms
    transfer latency p30 = 14363 ms
    transfer latency p40 = 22204 ms
    transfer latency p50 = 31483 ms
    transfer latency p60 = 41129 ms
    transfer latency p70 = 50849 ms
    transfer latency p80 = 60644 ms
    transfer latency p90 = 69450 ms
    transfer latency p100 = 76315 ms
    ```

### CODE REVIEW

File: src/lsm/compaction.zig [hunk 1]
```diff
@@ -483,18 +499,17 @@ pub fn CompactionType(
...
-                if (compaction.drop_tombstones) {
-                    compaction.copy_drop_tombstones();
-                } else {
-                    compaction.copy(.a);
+            {
+                var a_iterator = ArrayIterator{ .values = compaction.values_in[0] };
+                defer compaction.values_in[0] = a_iterator.values;
+
+                var b_iterator = ArrayIterator{ .values = compaction.values_in[1] };
+                defer compaction.values_in[1] = b_iterator.values;
```
Comment (sm2n) [inline src/lsm/compaction.zig:507]: Why is setting `values_in` necessary? Is it just to be cautious against aliasing? It doesn't look like it's used anywhere

File: src/lsm/compaction.zig [hunk 2]
```diff
@@ -483,18 +499,17 @@ pub fn CompactionType(
...
+
+                switch (compaction.context.table_info_a) {
+                    .immutable => |*it| compaction.compact_with(it, &b_iterator),
```
Comment (sm2n) [inline src/lsm/compaction.zig:510]: I think it'd be nice to have a comment about how and why we ignore a_iterator

File: src/lsm/compaction.zig [hunk 3]
```diff
@@ -508,122 +523,112 @@ pub fn CompactionType(
...
-            // Copy variables locally to ensure a tight loop.
-            const values_in_a = compaction.values_in[0];
             const values_out = compaction.table_builder.data_block_values();
-            var values_in_a_index: usize = 0;
             var values_out_index = compaction.table_builder.value_count;
+            defer compaction.table_builder.value_count = values_out_index;
+
+            // Specialize on ArrayIterator to do a memcpy when possible.
+            if (@TypeOf(value_iterator) == *ArrayIterator) {
+                if (!drop_tombstones) {
```
Comment (sm2n) [inline src/lsm/compaction.zig:555]: These if conditions can be merged with logical and

File: src/lsm/compaction.zig (hunk 1 above)
Comment (sm2n) [reply src/lsm/compaction.zig:507]: Oh, is that moved from https://github.com/tigerbeetledb/tigerbeetle/pull/769/files?diff=split&w=0#diff-525e7c0a1dfc2950de95e578bd84fa5d04ca25f5b2d28e43718916fe2f929a90L624 ?

**Comment by kprotty** [comment]
Will be covered by newer Object Cache changes.

**Pull request closed without merging**